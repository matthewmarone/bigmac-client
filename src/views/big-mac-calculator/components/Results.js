import React, { useState, useCallback } from "react";
import PropTypes from "prop-types";
import { makeStyles } from "@material-ui/core/styles";
import { Grid } from "@material-ui/core";
import { CurrencyForm, LocalResults, RandomResults } from "components";
import { getMaxNumOfGoodsInForiegnCountry, convertFunds } from "helpers";

const useStyles = makeStyles({
  root: {
    height: "100%",
  },
});

/**
 *
 * @param {*} props
 */
export const Results = (props) => {
  const classes = useStyles();

  const [localAmount, setLocalAmount] = useState(0);
  const {
    localCountry,
    localPrice,
    dollarPrice,
    dollarPPP,
    dollarPriceRandom,
    randomCountry,
  } = props;

  const localNumberOfBigMacs = Math.floor(localAmount / localPrice);

  const randomNumberOfBigMacs = getMaxNumOfGoodsInForiegnCountry(
    localAmount,
    localPrice,
    dollarPrice,
    dollarPriceRandom
  );

  const randomConvertedAmount = convertFunds(
    localAmount,
    dollarPrice,
    dollarPriceRandom
  );

  const handleLocalAmountChange = useCallback((amt) => {
    setLocalAmount(amt > 0 ? amt : 0);
  }, []);

  return (
    <Grid
      container
      spacing={3}
      direction="row"
      justify="space-evenly"
      alignItems="center"
      classes={{
        root: classes.root,
      }}
    >
      <Grid item xs={12}>
        <CurrencyForm
          country={localCountry}
          amount={localAmount}
          onAmountChange={handleLocalAmountChange}
        />
      </Grid>
      <Grid item xs={12}>
        {localAmount === 0 ? (
          ""
        ) : (
          <LocalResults
            country={localCountry}
            numOfBigMac={localNumberOfBigMacs}
            dollarPPP={dollarPPP}
          />
        )}
      </Grid>
      <Grid item xs={12}>
        {localAmount === 0 ? (
          ""
        ) : (
          <RandomResults
            country={randomCountry}
            amount={localAmount}
            amountConverted={randomConvertedAmount}
            numberOfBigMacs={randomNumberOfBigMacs}
          />
        )}
      </Grid>
    </Grid>
  );
};
Results.propTypes = {
  localCountry: PropTypes.string.isRequired,
  localPrice: PropTypes.number.isRequired,
  dollarPrice: PropTypes.number.isRequired,
  dollarPPP: PropTypes.number.isRequired,
  randomCountry: PropTypes.string.isRequired,
  dollarPriceRandom: PropTypes.number.isRequired,
};
