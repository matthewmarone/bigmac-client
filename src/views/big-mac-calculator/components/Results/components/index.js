export { default as LocalResults } from "./local-results";
export { default as CurrencyForm } from "./currency-form";
export { default as RandomResults } from "./random-results";
