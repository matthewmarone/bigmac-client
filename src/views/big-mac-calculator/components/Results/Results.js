import React, { useState, useCallback } from "react";
import PropTypes from "prop-types";
import { makeStyles } from "@material-ui/core/styles";
import { Grid } from "@material-ui/core";
import { CurrencyForm, LocalResults, RandomResults } from "./components";
import {
  getHowManyCanYouBuy,
  getHowManyYouCouldBuyInForiegnCountry,
  getRealitiveWorthAbroad,
} from "helpers";

const useStyles = makeStyles({
  root: {
    height: "100%",
  },
});

/**
 *
 * @param {*} props
 */
const Results = (props) => {
  const classes = useStyles();
  // Store users currency input
  const [localAmount, setLocalAmount] = useState(0);
  // Callback to handle currency input change
  const handleLocalAmountChange = useCallback((amt) => {
    // Only allow positive currency input
    setLocalAmount(!isNaN(amt) && amt > 0 ? amt : 0);
  }, []);
  const {
    localCountry,
    localPrice,
    dollarPrice,
    dollarPPP,
    dollarPriceRandom,
    randomCountry,
  } = props; // All these props are required as specified by PropTypes

  // Calc the number of big macs that can be bought in the current country
  const localNumberOfBigMacs = getHowManyCanYouBuy(localPrice, localAmount);

  // Calc the number of big macs that can be bought in a random country
  const randomNumberOfBigMacs = getHowManyYouCouldBuyInForiegnCountry(
    localAmount,
    localPrice,
    dollarPrice,
    dollarPriceRandom
  );

  // Cacl what the users local amount would "feel like" in the random country
  // based on the difference in the U.S. dollar price of the Big Mac between the
  // two countries
  const randomConvertedAmount = getRealitiveWorthAbroad(
    localAmount,
    dollarPrice,
    dollarPriceRandom
  );

  return (
    <Grid
      container
      spacing={3}
      direction="row"
      justify="space-evenly"
      alignItems="center"
      classes={{
        root: classes.root,
      }}
    >
      <Grid item xs={12}>
        {/* Display's top third of page */}
        <CurrencyForm
          country={localCountry}
          amount={localAmount}
          onAmountChange={handleLocalAmountChange}
        />
      </Grid>
      <Grid item xs={12}>
        {/* Display in middle of page only after user inputs amount */}
        {localAmount === 0 ? (
          ""
        ) : (
          <LocalResults
            country={localCountry}
            numOfBigMac={localNumberOfBigMacs}
            dollarPPP={dollarPPP}
          />
        )}
      </Grid>
      <Grid item xs={12}>
        {/* Display in bottom 3rd of page only after user inputs amount */}
        {localAmount === 0 ? (
          ""
        ) : (
          <RandomResults
            country={randomCountry}
            localCountry={localCountry}
            amount={localAmount}
            amountConverted={randomConvertedAmount}
            numberOfBigMacs={randomNumberOfBigMacs}
          />
        )}
      </Grid>
    </Grid>
  );
};
Results.propTypes = {
  localCountry: PropTypes.string.isRequired,
  localPrice: PropTypes.number.isRequired,
  dollarPrice: PropTypes.number.isRequired,
  dollarPPP: PropTypes.number.isRequired,
  randomCountry: PropTypes.string.isRequired,
  dollarPriceRandom: PropTypes.number.isRequired,
};

export default Results;
