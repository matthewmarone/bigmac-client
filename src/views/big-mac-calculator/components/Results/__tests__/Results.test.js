import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Results from "./../Results";
import { act } from "react-dom/test-utils";

test("Big Mac Calculator Resutls Component", async () => {
  // The necessary data to calculate, from big-mac-index.csv
  const props = {
    localCountry: "Argentina",
    localPrice: 33,
    dollarPrice: 2.389702554,
    dollarPPP: 6.693711968,
    randomCountry: "Australia",
    dollarPriceRandom: 3.743655,
  };
  const INPUT = "132.50"; // The users input

  await act(async () => {
    const { baseElement, getByTestId } = render(<Results {...props} />);
    const userInput = getByTestId("component-MoneyInput").querySelector(
      "input"
    );
    //Sim having the user input 23
    fireEvent.change(userInput, { target: { value: INPUT } });
    await new Promise((resolve) => setTimeout(resolve, 0)); // wait for response
    expect(baseElement).toMatchSnapshot();
  });
});
