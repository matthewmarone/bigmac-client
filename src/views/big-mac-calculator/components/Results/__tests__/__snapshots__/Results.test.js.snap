// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Big Mac Calculator Resutls Component 1`] = `
<body>
  <div>
    <div
      class="MuiGrid-root makeStyles-root-1 MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-align-items-xs-center MuiGrid-justify-xs-space-evenly"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <div>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorPrimary"
          >
            You are in 
            Argentina
            ! That's a fantastic place.
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorPrimary"
          >
            How much money in 
            Argentina
            's currency do you have?
          </p>
          <div
            class="MuiFormControl-root"
          >
            <label
              class="MuiFormLabel-root MuiFormLabel-colorSecondary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled Mui-focused Mui-focused"
              data-shrink="true"
              for="amountInput"
            >
              Amount
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorSecondary MuiInput-colorSecondary Mui-focused Mui-focused MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
              data-testid="component-MoneyInput"
            >
              <div
                class="MuiInputAdornment-root MuiInputAdornment-filled MuiInputAdornment-positionStart"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                >
                  $
                </p>
              </div>
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
                id="amountInput"
                type="number"
                value="0"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <div>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorPrimary"
          >
            Awesome! You could buy 
            4
             Big Macs in 
            Argentina
            .
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorPrimary"
          >
            Your Dollar Purchasing Parity (PPP) is 
            6.693711968
            .
          </p>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <div>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorPrimary"
          >
            But if you found yourself in 
            Australia
            , then you could buy
             
            2
             Big Macs with your 
            $132.50
            .
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorPrimary"
          >
            Likewise, your 
            $132.50
             from 
            Argentina
             feels more like
             
            $84.58
             in 
            Australia
            .
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
